posts/inside-rust/2019-11-23-const-if-match.md

```diff
@@ -0,0 +1,156 @@
+`||`, were already legal in a `const` or `static`. This was accomplished by
+translating them to their non-short-circuiting equivalents, `&` and `|`
+respectively. Enabling the feature gate will turn off this hack and make `&&`
+and `||` behave as you would expect.
+
+As a side-effect of these changes, the `assert` and `debug_assert` macros
+become usable in a const context if `#![feature(const_panic)]` is also
+enabled. However, the other assert macros (e.g., `assert_eq`,
+`debug_assert_ne`) remain forbidden, since they need to call `Debug::fmt` on
+their arguments.
+
+Also forbidden are looping constructs, `while`, `for`, and `loop`, which will
+be [feature-gated separately][52000], and the `?` operator, which calls
+`From::from` on the value inside the `Err` variant. The design for
+`const` trait methods is still being discussed.
+
+[52000]: https://github.com/rust-lang/rust/issues/52000
+
+## What's next?
+
+This change will allow a great number of standard library functions to be made
+`const`. You can help with this process! To get started, here's a [list of
+numeric functions][const-int] that can be constified with little effort.
+Conversion to a `const fn` requires two steps. First, `const` is added to a
+function definition along with a `#[rustc_const_unstable]` attribute. This
+allows nightly users to call it in a const context. Then, after a period of
+experimentation, the attribute is removed and the constness of that function is
+stabilized. See [#61635] for an example of the first step and [#64028] for an
+example of the second.
+
+Personally, I've looked forward to this feature for a long time, and I can't
+wait to start playing with it. If you feel the same, I would greatly
+appreciate if you tested the limits of this feature! Try to sneak `Cell`s and
+types with `Drop` impls into places they shouldn't be allowed, blow up the
+stack with poorly implemented recursive functions (see `gcd` above), and let
+us know if something goes horribly wrong.
+
+[const-int]: https://github.com/rust-lang/rust/issues/53718
+[#61635]: https://github.com/rust-lang/rust/issues/61635
+[#64028]: https://github.com/rust-lang/rust/pull/64028
+
+## What took you so long?
+
+[Miri], which rust uses under the hood for compile-time function evaluation,
+has been capable of this for a while now. However, rust needs to statically
+guarantee certain properties about variables in a `const`, such as whether
+they allow for interior mutability or whether they have a `Drop`
+implementation that needs to be called. For example, we must reject the
+following code since it would result in a `const` being mutable at runtime!
+
+[Miri]: https://github.com/rust-lang/miri
+
+```rust
+const CELL: &std::cell::Cell<i32> = &std::cell::Cell::new(42); // Not allowed...
+
+fn main() {
+    CELL.set(0);
+    println!("{}", CELL.get()); // otherwise this could print `0`!!!
+}
+```
+
+However, it is sometimes okay for a `const` to contain a *type* that allows
```

Comment:
It's also OK for constants to have their base value have interior mutability since each use site creates its own copy of it

Reply:
"a `const` to contain a reference to a *type* that..."?

Reply:
Yea that's good